host/docs/usrp2_card_burner.1
.TH "usrp2_card_burner" 1 "3.7.0" UHD "User Commands"
.SH NAME
usrp2_card_burner - USRP N-Series FPGA/Firmware Burner
.SH DESCRIPTION
Burn USRP2 firmware and FPGA images onto an SD card plugged into
the host machine.
.SH SYNOPSIS
.B  usrp2_card_burner [OPTIONS]
.SH OPTIONS
.IP "Raw device path:"
--addr=\fI"Address"\fR (default: 192.168.10.2)
.IP "This help information:"
--help
.IP "Firmware Filepath:"
--fw=\fI"filepath"\fR
.IP "FPGA Filepath:"
--fpga=\fI"filepath"\fR
.IP "List all USRP devices without burning"
--list
.IP "Override safety check"
--force
.SH EXAMPLES
.SS Standard burning procedure
.sp
usrp2_card_burner --dev=/dev/sdd1 --fw=usrp2_fw.bin --fpga=usrp2_fpga.bin
.ft
.fi
.SH SEE ALSO
UHD documentation:
.B http://files.ettus.com/manual/
.LP
GR-UHD documentation:
.B http://gnuradio.org/doc/doxygen/page_uhd.html
.LP
Other UHD programs:
.sp
uhd_images_downloader(1) usrp_n2xx_simple_net_burner(1) usrp_x3xx_fpga_burner(1)
.SH AUTHOR
This manual page was written by Nicholas Corgan
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2012,2014 Ettus Research LLC
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


CHANGELOG
Change Log for Releases
==============================

## 003.007.002
* Properly flushing PCIe chain on device to prevent stale data.
* Adjusted bus clock rate in FPGA to improve timing.
* Fixed issue where FPGA would fail to load FPGA image over PCIe.
* Fixed incompatibility issue with USRP2 FPGA burner utility.
* Fixed issue where ZPU would report empty NMEA strings from GPSDO.
* Updated some functions to no longer use deprecated Boost calls.
* Fixed issue where libusb would declare 'connection refused'.
* Fixed Windows library package naming.
* Updated documentation.
* Fixed reversed TX / RX colors in X300 LED indicators.
* DBSRX2 now works properly with X300 / X310.
* Updated include files to build on older distros of Linux.
* Fixed issue with 'item32' type converter from GNU Radio.
* Fixed issue where channels on B210 would occasionally not be phase-aligned.
* Fixed problem causing channels on B210 to swap between runs.
* Fixed issue in N-Series devices causing scaling error @ 50 MSps.
* The B200 / B210 now generates an internal PPS if none is provided.
* Improved performance on PCIe for X300.
* B200 / B210 properly reports clock rate (issue with float comparison).
* Fixed issue with filter selection in B200 that caused loss of RX power.
* Removed ref lock check from X300 where it wasn't necessary.

## 003.007.001
* Fixed issue with TVRX2 divider calculation.
* Fixed issue using calibration utilities on B-Side daughterboard in an X3xx.
* Replaced unsafe `sscanf` call in utilities.
* Properly initializing N-Series clock, fixing short transient on device boot-up.
* Improved `--help` output for a number of utilities & examples.
* Improved READMEs for directories.
* Fixed X3xx documentation with some clarifications.
* UHD will now tell you if you have requested a sample rate higher than the transport can deliver.
* Removed work-around necessary in RIO kernel module for zero-copy in PCIe for X3xx devices.
* Fixed issue where X3xx devices would lock-up when on networks with lots of traffic.
* The B2xx FX3 and AD9361 source code is now in UHD.
* Numerous B2xx stability improvements.
* Fixed includes for older OSes (e.g., Fedora 14).
* Fixed includes for older versions of Boost.
* Fixed PPS detection in X3xx with multiple time sources.
* Fixed overflow reporting for X3xx utilities.
* Fixed MTU / frame size detection for X3xx devices.
* Fixed B2xx filter chain causing wrong sample rate in some circumstances.

## 003.007.000
* Introduced USRP X300 and X310 support!
* Releasing a CHDR Dissector for Wireshark analysis
* Improved USRP B200 and B210 stability
* Introducing Integer-N tuning for WBX, SBX, CBX daughterboards
* Introducing support for 120 MHz versions of WBX, SBX, CBX
* Lots of new documentation
* New GPIO example for USRP X300
* Fixed threading bug in USRP B2xx code causing GQRX issue
* General UHD bug fixes & improvements

## 003.006.002
* Fixed bug in timed commands example.
* Improved compatibility for older versions of Boost.
* Fixed segmentation fault issue in converter code.
* Fixed binary-stripping for ARM builds.
* B2xx: Further improvements to code and images.
* Documentation links now installed to Start Menu in Windows.
* B2xx: More robust error checking in host code.
* B2xx: Improved error reporting.
* B2xx: Increase FPGA loading speed on USB 3
* B2xx: Re-programming FPGA without cycle now supported.
* B2xx: Larger RX SPP ceiling to support 16K transfers with larger RX FIFO in new FPGA image
* Updated internal READMEs and documentation.
* N-Series: netburner now accepts relative paths with '~'
* Completely re-written UHD Images Downloader, with numerous bug fixes and new features.

## 003.006.001
* B2xx: Fixed critical bug in 003.006.000 regarding USRP B2xx operation

## 003.006.000
* Many small fixes for bugs revealed by static analysis.
* Introduced support for the USRP B200 / B210

## 003.005.003

* E110:
** Fix FPGA Makefile build typo
* UHD:
** Fixed timespec irrational rate rounding
** Multichan streamer CPU utilization

## 003.005.002

* B100:
** Fix get send buffer timeout
* E1x0:
** Changes to bus timings for S issue
* USRP1:
** Restore broken EEPROM writing


## 003.005.001

* Misc:
** Fixes to images downloader
** Fixes to C++ net burner
** Added sleep(1) to query_gpsdo_sensors
* OSX:
** Fix for socket send code newer OS versions
* Changes from 3.4.5

## 003.005.000

* B100:
** Added timed commands feature
** Incremented FPGA compat number to 11.1
** Incremented firmware compat number to 3
* E1x0:
** Added timed commands feature
** Incremented FPGA compat number to 11.1
* USRP2/N2x0:
** Alternative stream destination on TX
** Incremented FPGA compat number to 10
* N2x0:
** Implemented timed-commands feature
** Implemented fast-commands feature
* SBX/WBX
** Tune with phase sync using timed-commands
* RFX series
** Added calibration utilities support
* General:
** SSE2 conversions for sc8 RX samples
** Added multi-threading to packet converters
** Added automatic images fetcher application

## 003.004.005

* XCVR2450:
** Added XCVR2450 rev2 support
* WBX:
** Added WBX rev4 support
* Misc:
** Fix to rx_samples_to_udp byte count
** stream command enums easy on SWIG

## 003.004.004

* Gen2
** Fix RX and TX DSP scalar adjustments
* B100/E1x0
** Fixed RX ADC IQ inversion swap
** Incremented FPGA compat number to 9.4

## 003.004.003

* USRP2/N2x0:
** Net burner parser fix (windows)
* XCVR2450:
** Fix to disable automatic LO offset on TX
* N2x0:
** Deal with misc exceptions in net burner
* E1x0:
** Changes to add reliability to bus state machine
* USRP1:
** Shutoff the DAC on transmit EOB flags
** Revert 1st nyquist zone DAC calculation

## 003.004.002

* USRP2/N2x0:
** Card and net burner language fixes
** Net burner python v3 code fix
** Net burner IPv6 interface fix
* E1x0:
** Fix for FPGA timing issue with GPMC input
** Incremented FPGA compat number to 9.2
* B100:
** Fix USB wrapper/buffer release race condition
* USRP1:

host/utils/usrp2_card_burner.py
#!/usr/bin/env python
#
# Copyright 2010-2011 Ettus Research LLC
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import platform
import tempfile
import subprocess
try:
    import urllib.request
except ImportError:
    import urllib
    urllib.request = urllib
import optparse
import math
import os
import re

########################################################################
# constants
########################################################################
SECTOR_SIZE = 512                 # bytes
MAX_FILE_SIZE =  1 * (2**20)      # maximum number of bytes we'll burn to a slot

FPGA_OFFSET = 0                   # offset in flash to fpga image
FIRMWARE_OFFSET = 1 * (2**20)     # offset in flash to firmware image

MAX_SD_CARD_SIZE = 2048e6         # bytes (any bigger is sdhc)

########################################################################
# helper functions
########################################################################
def command(*args):
    p = subprocess.Popen(
        args,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
    )
    ret = p.wait()
    verbose = p.stdout.read().decode('utf-8')
    if ret != 0: raise Exception(verbose)
    return verbose

def get_dd_path():
    if platform.system() == 'Windows':
        dd_path = os.path.join(os.path.dirname(__file__), 'dd.exe')
        if os.path.exists(dd_path): return dd_path
        dd_path = os.path.join(tempfile.gettempdir(), 'dd.exe')
        if not os.path.exists(dd_path):
            print('Downloading dd.exe to %s'%dd_path)
            dd_bin = urllib.request.urlopen('http://files.ettus.com/dd.exe').read()
            open(dd_path, 'wb').write(dd_bin)
        return dd_path
    return 'dd'

def int_ceil_div(num, den):
    return int(math.ceil(float(num)/float(den)))

def get_tmp_file():
    tmp = tempfile.mkstemp()
    os.close(tmp[0])
    return tmp[1]

########################################################################
# list possible devices
########################################################################
def get_raw_device_hints():
    ####################################################################
    # Platform Windows: parse the output of dd.exe --list
    ####################################################################
    if platform.system() == 'Windows':
        def extract_info_value(info, key):
            return info.split(key)[-1].split()[0]
        def get_info_list(output):
            in_info = False
            for line in output.splitlines():
                if line.startswith('\\\\'): in_info = True; info = ''
                elif in_info and not line.strip(): in_info = False; yield info
                if in_info: info += '\n'+line.strip()
        def is_info_valid(info):
            try:
                if 'link to' not in info: return False
                #handles two spellings of remov(e)able:
                if 'remov' not in info.lower(): return False
                if 'size is' in info and int(extract_info_value(info, 'size is')) > MAX_SD_CARD_SIZE: return False
            except: return False
            return True
        def extract_info_name(info):
            for key in ('Mounted on', 'link to'):
                if key in info: return extract_info_value(info, key)
            return info.splitlines()[0].strip()

        return sorted(set(map(extract_info_name, list(filter(is_info_valid, get_info_list(command(get_dd_path(), '--list')))))))

    ####################################################################
    # Platform Linux: parse procfs /proc/partitions
    ####################################################################
    if platform.system() == 'Linux':
        devs = list()
        for line in command('cat', '/proc/partitions').splitlines():
            try:
                major, minor, blocks, name = line.split()
                if not name[-1].isdigit() and int(minor) == 0: continue
                if int(blocks)*1024 > MAX_SD_CARD_SIZE: continue
            except: continue
            devs.append(os.path.join('/dev', name))

        return sorted(set(devs))

    ####################################################################
    # Platform Mac OS X: parse diskutil list and info commands
    ####################################################################
    if platform.system() == 'Darwin':
        devs = [d.split()[0] for d in [l for l in command('diskutil', 'list').splitlines() if l.startswith('/dev')]]
        def output_to_info(output):
            return dict([list(map(lambda x: x.strip(), pair.lower().split(':'))) for pair in [l for l in output.splitlines() if ':' in l]])
        def is_dev_valid(dev):
            info = output_to_info(command('diskutil', 'info', dev))
            try:
                if 'internal' in info and info['internal'] == 'yes': return False
                if 'ejectable' in info and info['ejectable'] == 'no': return False
                if 'total size' in info:
                    size_match = re.match('^.*\((\d+)\s*bytes\).*$', info['total size'])
                    if size_match and int(size_match.groups()[0]) > MAX_SD_CARD_SIZE: return False
            except: return False
            return True

        return sorted(set(filter(is_dev_valid, devs)))

    ####################################################################
    # Platform Others:
    ####################################################################
    return ()

########################################################################
# write and verify with dd
########################################################################
def verify_image(image_file, device_file, offset):
    #create a temporary file to store the readback image
    tmp_file = get_tmp_file()

    #read the image data
    img_data = open(image_file, 'rb').read()
    count = int_ceil_div(len(img_data), SECTOR_SIZE)

    #execute a dd subprocess
    verbose = command(
        get_dd_path(),
        "of=%s"%tmp_file,
        "if=%s"%device_file,
        "skip=%d"%(offset/SECTOR_SIZE),
        "bs=%d"%SECTOR_SIZE,
        "count=%d"%count,
    )

    #verfy the data
    tmp_data = open(tmp_file, 'rb').read(len(img_data))
    if img_data != tmp_data: return 'Verification Failed:\n%s'%verbose
    return 'Verification Passed:\n%s'%verbose

def write_image(image_file, device_file, offset):
    #create a temporary file to store the padded image
    tmp_file = get_tmp_file()

    #write the padded image data
    img_data = open(image_file, 'rb').read()
    count = int_ceil_div(len(img_data), SECTOR_SIZE)
    pad_len = SECTOR_SIZE*count - len(img_data)
    padding = bytes(b'\x00')*pad_len #zero-padding
    open(tmp_file, 'wb').write(img_data + padding)

    #execute a dd subprocess
    verbose = command(
        get_dd_path(),
        "if=%s"%tmp_file,
        "of=%s"%device_file,
        "seek=%d"%(offset/SECTOR_SIZE),
        "bs=%d"%SECTOR_SIZE,
        "count=%d"%count,
    )

    try: #exec the sync command (only works on linux)
        if platform.system() == 'Linux': command('sync')
    except: pass

    return verbose

def write_and_verify(image_file, device_file, offset):
    if os.path.getsize(image_file) > MAX_FILE_SIZE:
        raise Exception('Image file larger than %d bytes!'%MAX_FILE_SIZE)
    return '%s\n%s'%(
        write_image(
            image_file=image_file,
            device_file=device_file,
            offset=offset,
        ), verify_image(
            image_file=image_file,
            device_file=device_file,
            offset=offset,
        ),
    )

def burn_sd_card(dev, fw, fpga):
    verbose = ''
    if fw: verbose += 'Burn firmware image:\n%s\n'%write_and_verify(
        image_file=fw, device_file=dev, offset=FIRMWARE_OFFSET
    )
    if fpga: verbose += 'Burn fpga image:\n%s\n'%write_and_verify(
        image_file=fpga, device_file=dev, offset=FPGA_OFFSET
    )
    return verbose

########################################################################
# command line options
########################################################################
def get_options():
    parser = optparse.OptionParser()
    parser.add_option("--dev",  type="string",       help="raw device path",                default='')
    parser.add_option("--fw",   type="string",       help="firmware image path (optional)", default='')
    parser.add_option("--fpga", type="string",       help="fpga image path (optional)",     default='')
    parser.add_option("--list", action="store_true", help="list possible raw devices",      default=False)
    parser.add_option("--force", action="store_true", help="override safety check",         default=False)
    (options, args) = parser.parse_args()

    return options

########################################################################
# main
########################################################################
if __name__=='__main__':
    options = get_options()
    device_hints = get_raw_device_hints()
    show_listing = options.list

    if not show_listing and not options.force and options.dev and options.dev not in device_hints:
        print('The device "%s" was not in the list of possible raw devices.'%options.dev)
        print('The card burner application will now exit without burning your card.')
        print('To override this safety check, specify the --force option.\n')
        show_listing = True

    if show_listing:
        print('Possible raw devices:')
        print('  ' + '\n  '.join(device_hints))
        exit()

    if not options.dev: raise Exception('no raw device path specified')
    print(burn_sd_card(dev=options.dev, fw=options.fw, fpga=options.fpga))
